### ISSUE TITLE
backslash / forward slash issue in generated docker run command on remote containers running in WSL 2

### ISSUE THREAD

**Issue by jflam** [microsoft/vscode-remote-release#4789]
- VSCode Version: 1.54.3 (also repros on insiders VS Code)
- Local OS Version: Win 10 Insiders 21349.1000
- Remote OS Version: Win 10 Insiders 21349.1000
- Remote Extension/Connection Type: SSH/Docker/WSL
- Docker version 20.10.5, build 55c4c88

Steps to Reproduce:

1. This is my devcontainer.json (self-generated)

```js
{
    "name": "on localhost",
    "image": "pytorch-tutorials",
    "mounts": ["source=/var/run/docker.sock,target=/var/run/docker.sock,type=bind"],
    "extensions": [
        "ms-python.python",
        "ms-toolsai.jupyter",
        "ms-python.vscode-pylance"
    ],
}
```

2. When I launch VS Code in WSL 2 by typing `code .` in the directory containing the project and re-open using the remote-containers extension, I see this in the log files (dumped below):

`string=is not a valid Windows path`.

It's trying to send a `docker run` command to WSL 2 with **backslashes**. If I run the exact same command replacing backslashes with forward slashes it runs perfectly.

Everything was working fine 8-9 days ago according to my docker logs, so something changed recently to cause this regression.

Note that I'm running on Windows Insider builds so that I can test GPU integration. I don't have a non-insiders machine handy to see if this repros on retail Windows builds.

```
[2021-03-30T17:32:11.132Z] [PID 38100] [14 ms] Remote-Containers 0.163.2 in VS Code 1.54.3 (2b9aebd5354a3629c3aba0a5f5df49f43d6689f8).
[2021-03-30T17:32:11.592Z] [PID 38100] [474 ms] Start: Resolving Remote
[2021-03-30T17:32:11.593Z] [PID 38100] [475 ms] Setting up container for folder or workspace: \\wsl.localhost\Ubuntu-18.04\home\jlam\src\ez\src\pytorch-tutorials
[2021-03-30T17:32:11.597Z] [PID 38100] [478 ms] Start: Check Docker is running
[2021-03-30T17:32:11.597Z] [PID 38100] [479 ms] Start: Run: docker version --format {{.Server.APIVersion}}
[2021-03-30T17:32:12.044Z] [PID 38100] [926 ms] Server API version: 1.41
[2021-03-30T17:32:12.192Z] [PID 38100] [1074 ms] Start: Run: git rev-parse --show-cdup
[2021-03-30T17:32:12.291Z] [PID 38100] [1173 ms] Start: Run: docker ps -q -a --filter label=vsch.local.folder=\\wsl.localhost\Ubuntu-18.04\home\jlam\src\ez\src\pytorch-tutorials --filter label=vsch.quality=stable
[2021-03-30T17:32:12.754Z] [PID 38100] [1636 ms] Start: Run: docker events --format {{json .}} --filter event=start
[2021-03-30T17:32:12.774Z] [PID 38100] [1656 ms] Start: Starting container
[2021-03-30T17:32:12.775Z] [PID 38100] [1657 ms] Start: Run: docker run --sig-proxy=false -a STDOUT -a STDERR --mount type=bind,source=\\wsl.localhost\Ubuntu-18.04\home\jlam\src\ez\src\pytorch-tutorials,target=/workspaces/pytorch-tutorials,consistency=cached --mount source=/var/run/docker.sock,target=/var/run/docker.sock,type=bind --mount type=volume,src=vscode,dst=/vscode -l vsch.local.folder=\\wsl.localhost\Ubuntu-18.04\home\jlam\src\ez\src\pytorch-tutorials -l vsch.quality=stable -l vsch.remote.devPort=0 -u ezuser --gpus=all --ipc=host --entrypoint /bin/sh pytorch-tutorials -c echo Container started ; trap "exit 0" 15; while sleep 1 & wait $!; do :; done
[2021-03-30T17:32:12.942Z] [PID 38100] 
[2021-03-30T17:32:13.380Z] [PID 38100] docker: Error response from daemon: \\wsl.localhost\Ubuntu-18.04\home\jlam\src\e
z\src\pytorch-tutorials%!(EXTRA string=is not a valid Windows path).
See 'docker run --help'.
[2021-03-30T17:32:13.419Z] [PID 38100] [2301 ms] Start: Run: docker ps -q -a --filter label=vsch.local.folder=\\wsl.localhost\Ubuntu-18.04\home\jlam\src\ez\src\pytorch-tutorials --filter label=vsch.quality=stable
[2021-03-30T17:32:13.913Z] [PID 38100] [2795 ms] Command failed: docker run --sig-proxy=false -a STDOUT -a STDERR --mount type=bind,source=\\wsl.localhost\Ubuntu-18.04\home\jlam\src\ez\src\pytorch-tutorials,target=/workspaces/pytorch-tutorials,consistency=cached --mount source=/var/run/docker.sock,target=/var/run/docker.sock,type=bind --mount type=volume,src=vscode,dst=/vscode -l vsch.local.folder=\\wsl.localhost\Ubuntu-18.04\home\jlam\src\ez\src\pytorch-tutorials -l vsch.quality=stable -l vsch.remote.devPort=0 -u ezuser --gpus=all --ipc=host --entrypoint /bin/sh pytorch-tutorials -c echo Container started ; trap "exit 0" 15; while sleep 1 & wait $!; do :; done
```

Does this issue occur when you try this locally?: Yes
Does this issue occur when you try this locally and all extensions are disabled?: How would I repro if I disabled the remote-containers extension?

**Comment by chrmarti** [comment]
It uses the Docker CLI on Windows (not WSL). `\\wsl.localhost\` seems new, that used to be `\\wsl\`. Is there any documentation on this change? /cc @aeschli

**Comment by jflam** [comment]
No docs that I'm aware of. Would it help if I tried to hunt down someone in Windows for you?

So it seems like a recent Windows Insiders build broke me. Any ideas for a workaround that I can use in the meantime, like a secret setting I can inject into devcontainer.json so I can override the generated docker run command?

**Comment by jflam** [comment]
Some more experiments. This command opens an Explorer window using `cmd.exe`:

```start \\wsl.localhost\Ubuntu-18.04\home\jlam\src\ez\src```

So it seems like it's Docker is over-aggressively syntax checking `wsl.localhost`. Changing the `docker run` command to `\\wsl\localhost\...` also fails:

```docker run --sig-proxy=false -a STDOUT -a STDERR --mount type=bind,source=\\wsl\localhost\Ubuntu-18.04\home\jlam\src\ez\src\pytorch-tutorials,target=/workspaces/pytorch-tutorials,consistency=cached --mount source=/var/run/docker.sock,target=/var/run/docker.sock,type=bind --mount type=volume,src=vscode,dst=/vscode -l vsch.local.folder=\\wsl.localhost\Ubuntu-18.04\home\jlam\src\ez\src\pytorch-tutorials -l vsch.quality=stable -l vsch.remote.devPort=0 -u ezuser --gpus=all --ipc=host --entrypoint /bin/sh pytorch-tutorials -c echo Container started ; trap "exit 0" 15; while sleep 1 & wait $!; do :; done
docker: Error response from daemon: failed with: can't access specified distro mount service: stat /run/guest-services/distro-services/localhost.sock: no such file or directory.
See 'docker run --help'.
'wait' is not recognized as an internal or external command,
operable program or batch file.```

**Comment by jflam** [comment]
I opened [this issue](https://github.com/docker/for-win/issues/10786) on the Docker GH. I don't think this is a remote containers bug. Feel free to close if you agree. Thx!

**Comment by chrmarti** [comment]
`\\wsl\` stopped working for Docker's bind mounts with Docker for WSL2 on Windows. We since use the Docker CLI in WSL with Linux paths for setting up the bind mounts. We might simply need to handle `\\wsl.localhost\` the same way. I would first like to understand if this change in Windows is intentional and what other values we might see and what they mean.

**Comment by aeschli** [comment]
@chrmarti The WSL team has been discussing to add `\\wsl.localhost` as an alternative to `\\wsl$`.
The $ in the name is problematic for some applications, so they first added `\\wsl`, but are now thinking of switching to `\\wsl.localhost` and retiring `\\wsl` again.

**Comment by stephen-turner** [comment]
Is there any documentation or ticket about this change? Does it conform to the RFCs about the use of localhost? Or should I ask in microsoft/WSL?

**Comment by chrmarti** [comment]
@jflam Thanks for getting everyone involved in touch! I'll update Remote-Containers to also match `\\wsl.localhost\`. We use the Docker CLI on Linux when we detect a WSL path, I guess that should still work.

**Comment by mauve** [comment]
Hi this just broke for me in Insider Preview 21354, is there a release I can try?

**Comment by chrmarti** [comment]
Just released Remote-Containers 0.167.0 with the fix. This currently requires VS Code Insiders available here: https://code.visualstudio.com/insiders/

I haven't tested it, but the change is reasonably simple. Let me know if it works for you. Thanks.

**Comment by pbrit** [comment]
I just want to confirm that Remote-Containers 0.167.0 + VS Code Insiders fixes the issue.
Thank you, @chrmarti!

**Comment by jflam** [comment]
Can confirm that this works with vs code insiders. Also worth noting that [this fix in Docker desktop](https://github.com/docker/for-win/issues/10786#issuecomment-818466271) removes the need to have to run code insiders as well.

**Comment by chrmarti** [comment]
Note that without the fix here, Remote-Containers attempts (and with the fix from Docker succeeds) to mount the Windows path and uses the Docker CLI on Windows. With the fix in Remote-Containers 0.167.0, it uses the Docker CLI on Linux (as it used to with the older WSL paths).

There are subtle differences between these two, like which ssh-agent and git for credentials is being used.

**Comment by StefanScherer** [comment]
The issue is fixed in Docker Desktop 3.3.1 release. Docker Desktop now supports `\\wsl.localhost` pathes.
